### PR TITLE
feat: Provide easy prover setup

### PR DESCRIPTION
**Pull request by Artemka374**
## What ❔

Allow to run `zk_inception prover init` without `chain init`
Add docs for running provers and proving the batch.

## Why ❔

To provide easy way to spin up prover subsystem locally.

## Checklist

- [ ] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.
- [ ] Code has been formatted via `zk fmt` and `zk lint`.

### CODE REVIEW

File: zk_toolbox/crates/zk_inception/src/commands/chain/args/init.rs [hunk 1]
```diff
@@ -28,6 +28,8 @@ pub struct InitArgs {
     pub deploy_paymaster: Option<bool>,
     #[clap(long, help = MSG_L1_RPC_URL_HELP)]
     pub l1_rpc_url: Option<String>,
+    #[clap(long)]
+    pub copy_configs: Option<bool>,
```
Comment (Deniallugo) [inline zk_toolbox/crates/zk_inception/src/commands/chain/args/init.rs:32]: i can't imagine why, i won't it to be copied

File: zk_toolbox/crates/zk_inception/src/commands/prover/args/init.rs [hunk 2]
```diff
@@ -440,4 +463,49 @@ impl ProverInitArgs {
...
+            let DBNames { prover_name, .. } = generate_db_names(config);
+            let chain_name = config.name.clone();
+
+            let dont_drop = !PromptConfirm::new("Do you want to drop the database?").ask();
+
+            if PromptConfirm::new("Do you want to use default database?").ask() {
+                Some(ProverDatabaseConfig {
+                    database_config: DatabaseConfig::new(DATABASE_PROVER_URL.clone(), prover_name),
+                    dont_drop,
+                })
```
Comment (Deniallugo) [inline zk_toolbox/crates/zk_inception/src/commands/prover/args/init.rs:490]: Just add a default prompt for it

File: zk_toolbox/crates/zk_inception/src/commands/prover/args/init.rs [hunk 3]
```diff
@@ -440,4 +463,49 @@ impl ProverInitArgs {
...
+            } else {
+                let prover_db_url = Prompt::new(&msg_prover_db_url_prompt(&chain_name))
```
Comment (Deniallugo) [inline zk_toolbox/crates/zk_inception/src/commands/prover/args/init.rs:492]: it'd be nice to have an ability to provide all these arguments with a comand line

File: zk_toolbox/crates/zk_inception/src/commands/chain/args/init.rs (hunk 1 above)
Comment (Artemka374) [reply zk_toolbox/crates/zk_inception/src/commands/chain/args/init.rs:32]: Well, if they were already copied with prover init. Although, not a big deal I believe, I can remove it

File: zk_toolbox/crates/zk_inception/src/commands/prover/args/init.rs [hunk 4]
```diff
@@ -54,6 +60,20 @@ pub struct ProverInitArgs {
     #[serde(flatten)]
     pub setup_key_config: SetupKeyConfigTmp,
 
+    #[clap(long)]
+    pub copy_configs: Option<bool>,
```
Comment (Deniallugo) [inline zk_toolbox/crates/zk_inception/src/commands/prover/args/init.rs:64]: you could just check the existence and don't copy if it exists.

File: zk_toolbox/crates/zk_inception/src/commands/prover/args/init.rs [hunk 5]
```diff
@@ -440,4 +473,62 @@ impl ProverInitArgs {
...
+
+            if self.use_default.clone().unwrap_or_else(|| {
+                PromptConfirm::new(MSG_USE_DEFAULT_DATABASES_HELP)
+                    .default(true)
+                    .ask()
+            }) {
+                Some(ProverDatabaseConfig {
+                    database_config: DatabaseConfig::new(DATABASE_PROVER_URL.clone(), prover_name),
+                    dont_drop,
+                })
```
Comment (Deniallugo) [inline zk_toolbox/crates/zk_inception/src/commands/prover/args/init.rs:509]: I'd rather use promts with db name and so on. 
Use default needs, when you want to specify it from the console

File: prover/crates/bin/prover_cli/src/commands/insert_version.rs [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    let protocol_version = ProtocolVersionId::try_from(args.version)
+        .map_err(|_| anyhow::anyhow!("Invalid protocol version"))?;
+
+    let protocol_version_patch = VersionPatch(args.patch);
+
+    let snark_wrapper = H256::from_str(&args.snark_wrapper).unwrap_or_else(|_| {
+        panic!("Invalid snark wrapper hash");
+    });
+
+    conn.fri_protocol_versions_dal()
```
[the remark this answers is not in the document]
Comment (Artemka374) [reply prover/crates/bin/prover_cli/src/commands/insert_version.rs:42]: On the other hand - user will still need to do the same thing if he did it manually. So it's better to work with CL

File: prover/crates/bin/prover_cli/src/commands/insert_batch.rs [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+        .context("failed to build a prover_connection_pool")?;
+    let mut conn = connection.connection().await.unwrap();
+
+    let protocol_version = ProtocolVersionId::try_from(args.version)
+        .map_err(|_| anyhow::anyhow!("Invalid protocol version"))?;
+
+    let protocol_version_patch = VersionPatch(args.patch);
+
+    conn.fri_witness_generator_dal()
+        .save_witness_inputs(
```
[the remark this answers is not in the document]
Comment (Artemka374) [reply prover/crates/bin/prover_cli/src/commands/insert_batch.rs:35]: Just checked - no, we don't

File: prover/docs/05_proving_batch.md [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+Now, you can use `zk_inception` tool for setting up the env and running prover subsystem. Check the steps for
+installation in [this guide](../../zk_toolbox/README.md). And don't forget to install the prerequisites!
+
+Also, to assure that you are working with database securely, you need to use `prover_cli` tool. You can find the guide
+for installation [here](../crates/bin/prover_cli/README.md).
+
+## Initializing system
+
+After you have installed the tool, you can run the prover subsystem by running:
```
[the remark this answers is not in the document]
Comment (Artemka374) [reply prover/docs/05_proving_batch.md:21]: You are right! I think the point of zk_inception is to be run outside of zksync-era

File: prover/docs/05_proving_batch.md [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+
+```shell
+zk_inception ecosystem create
```
[the remark this answers is not in the document]
Comment (Artemka374) [reply prover/docs/05_proving_batch.md]: Right now it is fine to use all the default options, docs updated

File: prover/docs/05_proving_batch.md [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+### Getting data needed for proving
+
+At this step, we need to get the witness inputs data for the batch you want to prove. Database information now lives in
+input file, called `witness_inputs_<batch>.bin` generated by different core components).
+
+- If batch was produced by your system, the file is stored by prover gateway in GCS (or your choice of object storage --
+  check config). To access it from GCS (assuming you have access to the bucket), run:
+
+  ```shell
+  gsutil cp gs://your_bucket/witness_inputs/witness_inputs_<batch>.bin <path/to/era/prover/artifacts/witness_inputs>
```
[the remark this answers is not in the document]
Comment (Artemka374) [reply prover/docs/05_proving_batch.md:52]: Why should it? Won't it create the directory and put the artifacts there?

File: prover/docs/05_proving_batch.md [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+  the batch number as a query data to get the data for specific batch, otherwise, you will receive latest data for the
+  batch, that was already proven.
+
+### Preparing database
+
+After you have the data, you need to prepare the system to run the batch. So, database needs to know about the batch and
+the protocol version it should use. You can do that with running
+
+```shell
+zk_inception prover protocol-version
```
Comment (Deniallugo) [inline prover/docs/05_proving_batch.md:80]: could you move it zk supervisor? it'd make more sense there. i'd prefer to keep inception only for initialization

File: zk_toolbox/crates/zk_inception/src/commands/prover/protocol_version.rs [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+    ));
+
+    Ok(())
+}
+
+async fn get_protocol_version(shell: &Shell, link_to_code: &Path) -> anyhow::Result<String> {
+    let path = link_to_code.join("prover/crates/bin/prover_version/Cargo.toml");
+    let protocol_version = cmd!(
+        shell,
+        "cargo +nightly-2024-08-01 run --manifest-path {path}"
```
Comment (Deniallugo) [inline zk_toolbox/crates/zk_inception/src/commands/prover/protocol_version.rs:25]: why do you need to specify the version? 
we will take the version from `rust-toolchain`

Comment (Deniallugo) [reply zk_toolbox/crates/zk_inception/src/commands/prover/protocol_version.rs:25]: and why do you need to specify the whole manifest? 
seems it's easier to `push_dir('prover')` and run `cargo run --bin prover_version`

File: zk_toolbox/crates/zk_supervisor/src/main.rs [hunk 13]
```diff
@@ -103,6 +105,7 @@ async fn run_subcommand(args: Supervisor, shell: &Shell) -> anyhow::Result<()> {
         }
         SupervisorSubcommands::Lint(args) => commands::lint::run(shell, args)?,
         SupervisorSubcommands::Fmt(args) => commands::fmt::run(shell.clone(), args).await?,
+        SupervisorSubcommands::ProtocolVersion => commands::protocol_version::run(shell).await?,
```
Comment (Deniallugo) [inline zk_toolbox/crates/zk_supervisor/src/main.rs:108]: probably prover protocol version? or it's a universal?
should we add a new layer?

Comment (Artemka374) [reply zk_toolbox/crates/zk_supervisor/src/main.rs:108]: I mean... it's hard to say. It's more related to provers as they are using it, so yeah, I guess we can call it prover protocol version

Comment (Artemka374) [reply zk_toolbox/crates/zk_supervisor/src/main.rs:108]: Just that I'm not sure that it needs a new layer, I don't have idea what else can be used for provers there. We can change the name to prover-version

Comment (Deniallugo) [reply zk_toolbox/crates/zk_supervisor/src/main.rs]: let's keep it prover version and later on will change it

**Review by Deniallugo — APPROVED** [review]

**Pull request merged**